storage/storage_image.go

```diff
@@ -0,0 +1,569 @@
+			return err
+		}
+		logrus.Debugf("saved image metadata %q", string(metadata))
+	}
+	return nil
+}
+
+func (s *storageImageDestination) SupportedManifestMIMETypes() []string {
+	return nil
+}
+
+func (s *storageImageDestination) PutManifest(manifest []byte) error {
+	s.Manifest = make([]byte, len(manifest))
+	copy(s.Manifest, manifest)
+	return nil
+}
+
+// SupportsSignatures returns an error if we can't expect GetSignatures() to
+// return data that was previously supplied to PutSignatures().
+func (s *storageImageDestination) SupportsSignatures() error {
+	return nil
+}
+
+func (s *storageImageDestination) PutSignatures(signatures [][]byte) error {
+	sizes := []int{}
+	sigblob := []byte{}
+	for _, sig := range signatures {
+		sizes = append(sizes, len(sig))
+		newblob := make([]byte, len(sigblob)+len(sig))
+		copy(newblob, sigblob)
+		copy(newblob[len(sigblob):], sig)
+		sigblob = newblob
+	}
+	s.Signatures = sigblob
+	s.SignatureSizes = sizes
+	return nil
+}
+
+func (s *storageImageSource) GetBlob(digest string) (rc io.ReadCloser, n int64, err error) {
+	rc, n, _, err = s.GetBlobAndLayerID(digest)
+	return rc, n, err
+}
+
+func (s *storageImageSource) GetBlobAndLayerID(digest string) (rc io.ReadCloser, n int64, layerID string, err error) {
+	err = ddigest.Digest(digest).Validate()
+	if err != nil {
+		return nil, -1, "", err
+	}
+	if layerList, ok := s.Layers[digest]; !ok || len(layerList) < 1 {
+		b, err := s.imageRef.transport.store.GetImageBigData(s.ID, digest)
+		if err != nil {
+			return nil, -1, "", err
+		}
+		r := bytes.NewReader(b)
+		logrus.Debugf("exporting opaque data as blob %q", digest)
+		return ioutil.NopCloser(r), int64(r.Len()), "", nil
+	}
+	// If the blob was "put" more than once, we have multiple layer IDs
+	// which should all produce the same diff.  For the sake of tests that
+	// want to make sure we created different layers each time the blob was
+	// "put", though, cycle through the layers.
+	layerList := s.Layers[digest]
+	position, ok := s.LayerPosition[digest]
+	if !ok {
+		position = 0
+	}
+	s.LayerPosition[digest] = (position + 1) % len(layerList)
+	logrus.Debugf("exporting filesystem layer %q for blob %q", layerList[position], digest)
+	rc, n, err = s.diffLayer(layerList[position])
+	return rc, n, layerList[position], err
+}
+
+func (s *storageImageSource) diffLayer(layerID string) (rc io.ReadCloser, n int64, err error) {
+	layer, err := s.imageRef.transport.store.GetLayer(layerID)
+	if err != nil {
+		return nil, -1, err
+	}
+	layerMeta := storageLayerMetadata{
+		CompressedSize: -1,
+	}
+	if layer.Metadata != "" {
+		if err := json.Unmarshal([]byte(layer.Metadata), &layerMeta); err != nil {
+			logrus.Errorf("error decoding metadata for layer %q: %v", layerID, err)
+			return nil, -1, err
+		}
+	}
+	if layerMeta.CompressedSize <= 0 {
+		n, err = s.imageRef.transport.store.DiffSize("", layer.ID)
+		if err != nil {
+			return nil, -1, err
+		}
+	} else {
+		n = layerMeta.CompressedSize
```

Comment:
> We've inherited an implicit assumption that gzip-compressed blobs are compressed using `compress/gzip` at the DefaultCompression level

That is, sadly, not enough to guarantee that the compressed bitstream will be exactly reconstructed from the unchanged input over golang version changes, #157 .

> We could discard the compression instead, but that would break things elsewhere.

Ah, I _think_ I see what the issue is:

As far as individual blobs go, `GetBlob` really doesn’t need to compress the data if the native format is not compressed; `copy.Image` will compress the blob when uploading to a remote registry if that doesn’t break anything, and will not bother when using a local destination like `docker-daemon:`. (But if `containers/storage` prefers to hand out compressed output for its own reasons, that’s fine.)

The issue, though, is with the manifest; if `GetManifest` returns exactly the original unmodified manifest, that manifest contains digests and sizes of the original compressed data, then `GetBlob` needs to return a blob exactly matching that. OTOH, when compression is involved, we really can’t guarantee that (#157).

(And, to some extent, there is another design issue that containers/image is always identifying layers by blob digests, which doesn’t really work for to-be-compressed layers where the digest is unknown until the layer blob is created. But that’s easier to fix than the rest.)

Is that what we are struggling with here?

---

If so, then we currently really can’t deal with that, because `copy.go:copyLayers` always verifies the stream from `GetBlob` against the digest. So, yeah, just hoping for the best is the only way to get things working right now.

Another possibility would be to teach `copyLayers` that some digests are not meant to match; in that case, `copyLayers` already has the infrastructure to collect the _true_ digests and sizes uploaded to the destination, and to edit the manifest to use them. Of course changing the manifest this way completely breaks signatures.

Right now `docker-daemon:` does not support signatures so these issues are not present in there, so containers/storage is the first to hit these issues. Eventually, for https://trello.com/c/5WHV80NF/96-8-signing-container-image-tracking-signing , we will need to figure out how to deal with matching/applying/verifying signatures when we can only preserve the layer `DiffID` and not the digest of the compressed blob.

---

On balance, I think `GetBlob` from containers/storage should return either the true size of the returned output stream, or `-1` if that value is too expensive to compute; the digest/size mismatch issues are real but they will manifest at the digest verification / manifest editing level, and while we probably have to keep “hoping for the best” at _those_ levels, at least for a little while, I can’t see that returning the size we wish for instead of the real size from `GetBlob` helps.